tests/Router/Middleware/NotFoundTest.php

```diff
@@ -0,0 +1,197 @@
+        $this->assertTrue(!isset($result->nextCalled), "'next' was called");
+        
+        if (is_callable($notAllowed)) {
+            $this->assertTrue($result->notAllowedCalled, "'Not allowed' callback was not called");
+        } elseif (!$notAllowed && is_callable($notFound)) {
+            $this->assertTrue($result->notFoundCalled, "'Not found' callback was not called");
+        }
+    }
+
+    /**
+     * Text that 'next' callback is invoked when route is found
+     *
+     * @dataProvider invokeProvider
+     * @param callback|int $notFound 
+     * @param callback|int $notAllowed 
+     * @param callback $next
+     */
+    public function testInvokeNext($notFound, $notAllowed, $next)
+    {
+        if (!$next) return $this->skipTest();
+
+        $routes = $this->getRoutes();
+        $middleware = new NotFound($routes, $notFound, $notAllowed);
+        list($request, $response) = $this->getRequests('/foo', 'GET');
+
+        $result = $middleware($request, $response, $next);
+
+        $this->assertEquals(get_class($response), get_class($result), "Result must be an instance of 'ResponseInterface'");
+        $this->assertTrue($result->nextCalled, "'next' waas not called");
+    }
+
+    /**
+     * Provide data for testing invoke method
+     */
+    public function invokeProvider()
+    {
+        $callbacks = [];
+        foreach (['notFound', 'notAllowed', 'next'] as $type) {
+            $var = $type . 'Called';
+
+            $callbacks[$type] = function($request, $response) use ($var) {
+                $response->$var = true;
+                return $response;
+            };
+        }
+
+        return [
+            [404, 405, $callbacks['next']],
+            [404, 405, null],
+            [404, null, $callbacks['next']],
+            [404, null, null],
+            [$callbacks['notFound'], $callbacks['notAllowed'], $callbacks['next']],
+            [$callbacks['notFound'], $callbacks['notAllowed'], null],
+            [$callbacks['notFound'], null, $callbacks['next']],
+            [$callbacks['notFound'], null, null]
+        ];
+    }
+
+    /**
+     * Expect that response is set to simple deny answer
+     *
+     * @param ResponseInterface $response
+     * @param int $code 
+     */
+    public function expectSimpleDeny(ResponseInterface $response, $code)
+    {
+        $stream = $this->createMock(StreamInterface::class);
+        $stream->expects($this->once())->method('rewind');
+        $stream->expects($this->once())->method('write')->with($this->equalTo('Not Found'));
+
+        $response->method('getBody')->will($this->returnValue($stream));
+        $response->expects($this->once())->method('withBody')->with($this->equalTo($stream))->will($this->returnSelf());
+        $response->expects($this->once())->method('withStatus')->with($this->equalTo($code), $this->equalTo('Not Found'))->will($this->returnSelf());
+    }
+
```

Comment:
Please create tests for the case where
* the route **is** found
* the route isn't found at all
* the route isn't found, but exists for another method